File: backend/src/main/kotlin/nl/stevenbontenbal/chorister/model/entities/Invite.kt
package nl.stevenbontenbal.chorister.model.entities

import nl.stevenbontenbal.chorister.interfaces.ChoirOwnedEntity
import java.time.LocalDateTime
import java.util.*
import jakarta.persistence.*

@Entity
class Invite(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,
    var email: String?,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "CHOIR_ID")
    override var choir: Choir?,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "INVITED_BY_ID")
    var invitedBy: User?,
    var token: String = UUID.randomUUID().toString(),
    var createdDate: LocalDateTime = LocalDateTime.now(),
    var acceptedDate: LocalDateTime? = null,
    var expired: Boolean = false,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "USER_ID")
    var user: User? = null
) : ChoirOwnedEntity {
    fun isAccepted(): Boolean = acceptedDate != null && user != null

    companion object
}

File: frontend/src/services/apiTypes/invite.ts
import { ApiEntityIn, Identifiable } from "."
import { Choir, toDomainChoir } from "./choir"
import { User } from "./user"
import { Invite as DomainInvite } from "@/entities/invite"

export interface Invite {
    email?: string
    token: string
    choir: Choir
}

export interface SingleInvite extends Invite, Identifiable, ApiEntityIn {
    id: number
    email: string
    invitedBy: User
    token: string
    createdDate: Date
    acceptedDate: Date
    expired: Boolean
    choir: Choir
}

export function toDomainInvite(invite: Invite): DomainInvite {
    return {
        token: invite.token,
        choir: toDomainChoir(invite.choir)
    }
}
